File: tests/solutions_cpp/merge_k_sorted_lists.cpp
#include "json_helper.h"

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        // Min-heap: pair<value, index> (index breaks ties); store nodes separately
        using Item = tuple<int, int, ListNode*>;
        priority_queue<Item, vector<Item>, greater<Item>> pq;
        for (int i = 0; i < (int)lists.size(); i++) {
            if (lists[i]) pq.push({lists[i]->val, i, lists[i]});
        }
        ListNode dummy(0);
        ListNode* curr = &dummy;
        while (!pq.empty()) {
            auto [val, i, node] = pq.top(); pq.pop();
            curr->next = node;
            curr = curr->next;
            if (node->next) pq.push({node->next->val, i, node->next});
        }
        curr->next = nullptr;
        return dummy.next;
    }
};

int main() {
    auto args = parse_args();
    Solution sol;
    auto lists = build_lists(args[0]);
    auto result = sol.mergeKLists(lists);
    cout << list_to_json(result) << endl;
    return 0;
}

File: tests/solutions_js/merge_k_sorted_lists.js
class Solution {
    mergeKLists(lists) {
        if (!lists || lists.length === 0) return null;

        while (lists.length > 1) {
            const merged = [];
            for (let i = 0; i < lists.length; i += 2) {
                const l1 = lists[i];
                const l2 = i + 1 < lists.length ? lists[i + 1] : null;
                merged.push(this._mergeTwoLists(l1, l2));
            }
            lists = merged;
        }

        return lists[0];
    }

    _mergeTwoLists(l1, l2) {
        const dummy = new ListNode(0);
        let current = dummy;

        while (l1 && l2) {
            if (l1.val <= l2.val) {
                current.next = l1;
                l1 = l1.next;
            } else {
                current.next = l2;
                l2 = l2.next;
            }
            current = current.next;
        }

        current.next = l1 ? l1 : l2;
        return dummy.next;
    }
}
